projects/packages/publicize/src/class-publicize-base.php

```diff
@@ -1608,17 +1615,154 @@ public function get_social_opengraph_image( $post_id ) {
+			array(
+				'w' => $width,
+				'h' => $height,
+			)
+		);
+	}
+
+	/**
+	 * This function runs the image through Proton to compress it, and also scales it down if needed.
+	 *
+	 * @param string $url Image URL.
+	 * @param int    $width Image width.
+	 * @param int    $height Image height.
+	 * @return array The compressed image data.
+	 */
+	public function compress_and_scale_og_image( $url, $width, $height ) {
+		// If the dimensions are fine we just run it through Proton to compress it.
+		if ( 1200 >= $width && 1200 >= $height ) {
+			return array(
+				'url'    => $this->get_resized_image_url( $url, $width, $height ),
+				'width'  => $width,
+				'height' => $height,
+			);
+		}
+
+		if ( $height > $width ) {
+			// Portrait.
+			$width  = 1200 * $width / $height;
+			$height = 1200;
+		} else {
+			// Landscape.
+			$height = 1200 * $height / $width;
+			$width  = 1200;
+		}
+
+		return array(
+			'url'    => $this->get_resized_image_url( $url, $width, $height ),
+			'width'  => $width,
+			'height' => $height,
+		);
+	}
+
+	/**
+	 * Reduce the filesize of an image by reducing the dimensions. Uses Photon.
+	 * Returns null if the image cannot be reduced enough.
+	 *
+	 * @param string $url Image URL.
+	 * @param int    $width Image width.
+	 * @param int    $height Image height.
+	 * @param int    $filesize Image filesize.
+	 * @param int    $tries Number of times to try reducing the image size. Default is 5.
+	 * @return array|null
+	 */
+	public function reduce_file_size( $url, $width, $height, $filesize, $tries = 5 ) {
+		while ( $tries > 0 && $filesize > self::OG_IMAGE_MAX_FILESIZE ) {
+			$width   *= 0.75;
+			$height  *= 0.75;
+			$url      = $this->get_resized_image_url( $url, $width, $height );
+			$filesize = $this->get_remote_filesize( $url );
+			--$tries;
+		}
+
+		// If the image is still too large, we failed.
+		if ( $filesize > self::OG_IMAGE_MAX_FILESIZE ) {
+			// TODO: Track this to see if conversion failed.
+			return null;
+		}
+
+		return array(
+			'url'    => $url,
+			'width'  => $width,
+			'height' => $height,
+		);
+	}
+
+	/**
+	 * Hooks into jetpack_open_graph_tags to add the Jetpack Social images to the OpenGraph tags,
+	 * or to make the Jetpack open graph images pass restrictions.
 	 *
 	 * @param array $tags Current tags.
 	 */
-	public function add_jetpack_social_og_images( $tags ) {
-		$opengraph_image = $this->get_social_opengraph_image( get_the_ID() );
+	public function jetpack_social_open_graph_filter( $tags ) {
+		$social_opengraph_image = $this->get_social_opengraph_image( get_the_ID() );
+
+		if ( ! empty( $social_opengraph_image ) ) {
+			return $this->add_jetpack_social_og_image( $tags, $social_opengraph_image );
+		}
+
+		// If we do not have a SIG or attached image, but we have an image in post body
+		// we need to check that the image is not too large for the social sites.
+		if ( ! empty( $tags['og:image'] ) && ! empty( $tags['og:image:width'] ) && ! empty( $tags['og:image:height'] ) ) {
```

Comment:
We could remove a level of indentation if we invert this logic and return `$tags`